subprojects/build-cache/src/main/java/org/gradle/caching/local/internal/H2BuildCacheServiceFactory.java

```diff
@@ -63,11 +81,23 @@ public BuildCacheService createBuildCacheService(DirectoryBuildCache configurati
         checkDirectory(target);
 
         int removeUnusedEntriesAfterDays = configuration.getRemoveUnusedEntriesAfterDays();
-        describer.type(H2_BUILD_CACHE_TYPE).
-            config("location", target.getAbsolutePath());
+        describer.type(H2_BUILD_CACHE_TYPE)
+            .config("location", target.getAbsolutePath())
+            .config("removeUnusedEntriesAfter", removeUnusedEntriesAfterDays + " days");
+        H2BuildCacheService h2Service = new H2BuildCacheService(target.toPath(), parallelismConfiguration.getMaxWorkerCount(), removeUnusedEntriesAfterDays, Time.clock());
+        // TODO: H2Cache could be provided by PersistentCache
+        // For now PersistentCache is used just so we can reuse cleanup infrastructure, but in future H2Cache could be provided by PersistentCache
+        Function<HasCleanupAction, PersistentCache> persistentCacheFactory = buildCacheService -> unscopedCacheBuilderFactory
+            .cache(target)
+            .withCleanupStrategy(createCacheCleanupStrategy((cleanableStore, progressMonitor) -> buildCacheService.cleanup()))
+            .withDisplayName("Build cache NG")
+            .withLockOptions(mode(None))
```

Comment:
💭 When we go into production, let's revisit this. We could fold `LockOnDemandCrossProcessBuildCacheService` into the existing locking tools, i.e. it should be possible to pass "open" and "close" operations to `LockOptionsBuilder` when it's in `OnDemand` mode. But this can wait.

Perhaps adding a TODO here would be good.

Reply:
Hmm, checking the code of `LockOnDemandCrossProcessBuildCacheService`, can we just use PersistentCache to do the locking? When building PersistentCache we set `.withLockOptions(mode(OnDemand))` and then in `LockOnDemandCrossProcessBuildCacheService` we do `persistentCache.withFileLock(() -> ...)`.

Could that work? Or we need some additional locking here, e.g. I see we have `withOwnershipNow` methods?